cli/flox/src/commands/edit.rs

```diff
@@ -130,25 +165,55 @@ impl Edit {
     // instead of just environment is a pain
     async fn edit_manifest(
         flox: &Flox,
-        environment: &mut Box<dyn Environment>,
-        active_environment: UninitializedEnvironment,
-        description: String,
+        environment: &mut ConcreteEnvironment,
         contents: Option<String>,
     ) -> Result<()> {
-        match maybe_migrate_environment_to_v1(flox, environment, &description).await {
+        // TODO: we have various functionality spread across
+        // UninitializedEnvironment, ConcreteEnvironment, and
+        // Environment.
+        // UninitializedEnvironment is used to compare to what
+        // environments are active.
+        // description can't currently be derived from an Environment
+        // but is used for messages.
+        // Environment is what we'll actually use to perform the edit.
+
+        if let ConcreteEnvironment::Managed(ref environment) = environment {
+            if environment.has_local_changes(flox)? && contents.is_none() {
+                bail!(ManagedEnvironmentError::CheckoutOutOfSync)
+            }
+        };
+
+        match maybe_migrate_environment_to_v1(flox, environment).await {
             Ok(_) => (),
             e @ Err(MigrationError::MigrationCancelled) => e?,
+
             // If the user said they wanted an upgrade and it failed, print why but don't fail
-            Err(MigrationError::ConfirmedUpgradeFailed(environment_error)) => {
-                // TODO: this could probably benefit from some newlines
-                message::warning(display_chain(&environment_error));
+            // [CoreEnvironmentError::LockForMigration] and [CoreEnvironmentError::MigrateManifest]
+            // are handled separately to avoid suggesting the use of `flox edit` within `flox edit`.
+            Err(MigrationError::ConfirmedUpgradeFailed(EnvironmentError::Core(
+                CoreEnvironmentError::LockForMigration(err),
+            ))) => {
+                message::warning(format_core_error(&err));
```

Comment:
nonblocking cause I want to get this merged: doesn't `format_core_error` for a `LockForMigration` error still print `Use 'flox edit' to resolve errors and then try again.`?

Shouldn't we have
```
Could not automatically migrate manifest to version 1:
Failed to create version 1 lock:

{err}
```

Reply:
Only if it's a CoreEnvironmentError::LockForMigration(Box<CireEnvironmentError::{LockForMigration,MigrateManifest}>)

Not sure if that is even possible but it's for sure a bit weird due to this error being recursive now